File: backend/api/schemas/backtest.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class BacktestRequest(BaseModel):
    ticker: str
    train_start: date
    train_end: date
    test_start: date
    test_end: date
    horizon_days: int
    n_sims: int
    models: list[str] | None = None
    window_step_days: int = 5


class BacktestJobResponse(BaseModel):
    job_id: str
    status: str  # "queued" | "running" | "done" | "failed"


class BacktestResultResponse(BaseModel):
    job_id: str
    status: str
    run_id: str | None = None
    rankings: list[str] | None = None
    error: str | None = None
    # Real progress, not a client-side timer: `stage` is one of
    # "downloading_prices" | "detecting_regime" | "running_backtests" |
    # "aggregating_results" | "done"; `progress` is (completed, total)
    # windows scored, only populated during "running_backtests".
    stage: str | None = None
    progress: tuple[int, int] | None = None


class CancelJobResponse(BaseModel):
    job_id: str
    cancelled: bool  # False if the job had already finished (or never existed)


class BacktestHistoryResponse(BaseModel):
    ticker: str
    run_ids: list[str]


class PerformanceRow(BaseModel):
    regime: str
    model_name: str
    n_observations: float
    mean_mae: float
    mean_rmse: float
    mean_crps: float
    coverage_90: float
    coverage_95: float
    directional_accuracy: float
    mean_var_violations: float
    mean_es_95: float
    source: str


class PerformanceTableResponse(BaseModel):
    ticker: str
    used_fallback: bool  # True if no per-ticker history yet; showing the universal prior instead
    rows: list[PerformanceRow]


class BacktestWindowPoint(BaseModel):
    window_origin: date
    actual_price: float
    forecast_median_price: float
    forecast_p5_price: float
    forecast_p95_price: float
    regime: str | None = None


class BacktestDetailResponse(BaseModel):
    ticker: str
    run_id: str
    model_name: str
    model_display_name: str
    available_models: list[str]
    points: list[BacktestWindowPoint]
    mean_absolute_pct_error: float
    coverage_90: float
    directional_accuracy: float

File: shared/types/backtest.ts
// Mirrors backend/api/schemas/backtest.py

export interface BacktestRequest {
  ticker: string;
  train_start: string;
  train_end: string;
  test_start: string;
  test_end: string;
  horizon_days: number;
  n_sims: number;
  models?: string[] | null;
  window_step_days?: number;
}

export type BacktestJobStatus = "queued" | "running" | "done" | "failed" | "cancelled";

// Mirrors the `stage` values `BacktestEngine.run` reports via `on_progress`.
export type BacktestStage =
  | "downloading_prices"
  | "detecting_regime"
  | "running_backtests"
  | "aggregating_results"
  | "done";

export interface BacktestJobResponse {
  job_id: string;
  status: BacktestJobStatus;
}

export interface BacktestResultResponse {
  job_id: string;
  status: BacktestJobStatus;
  run_id: string | null;
  rankings: string[] | null;
  error: string | null;
  stage: BacktestStage | null;
  progress: [number, number] | null;
}

export interface CancelJobResponse {
  job_id: string;
  cancelled: boolean;
}

export interface BacktestHistoryResponse {
  ticker: string;
  run_ids: string[];
}

export interface PerformanceRow {
  regime: string;
  model_name: string;
  n_observations: number;
  mean_mae: number;
  mean_rmse: number;
  mean_crps: number;
  coverage_90: number;
  coverage_95: number;
  directional_accuracy: number;
  mean_var_violations: number;
  mean_es_95: number;
  source: string;
}

export interface PerformanceTableResponse {
  ticker: string;
  used_fallback: boolean;
  rows: PerformanceRow[];
}

export interface BacktestWindowPoint {
  window_origin: string;
  actual_price: number;
  forecast_median_price: number;
  forecast_p5_price: number;
  forecast_p95_price: number;
  regime: string | null;
}

export interface BacktestDetailResponse {
  ticker: string;
  run_id: string;
  model_name: string;
  model_display_name: string;
  available_models: string[];
  points: BacktestWindowPoint[];
  mean_absolute_pct_error: number;
  coverage_90: number;
  directional_accuracy: number;
}
